### ISSUE TITLE
Zero-lenth assertions?

### ISSUE THREAD

**Issue by introrse** [Hardmath123/nearley#89]
I'm loving nearley, thank you for building it!  This is vastly better than LR/LL parsing and I'm astonished LR/LL still garners so much attention given the limitations.  

A question: Is there a way to encode zero-length assertions or otherwise control ambiguity when one nonterminal is an abbreviation of another?  

For instance, consider a language of  'a', 'b', and 'ab' tokens where ab should be matchedinstead of 'a' followed by 'b'.  Given the rule
    tokens -> ("ab" | "a" | "b"):*

"aabab" would ideally parse to ["ab" | "a" | "b"]

Instead, you get 5 matches breaking up the "ab" tokens differently:
  [[["a"],["ab"],["ab"]]]
  [[["a"],["a"],["b"],["ab"]]]
  [[["a"],["a"],["ba"],["b"]]]
  [[["a"],["ab"],["a"],["b"]]]
  [[["a"],["a"],["b"],["a"],["b"]]]

I can get the ideal output if I write a grammar where a standalone "b" is not allowed.  But what if you need that?  (e.g., if you're trying to parse "elseif", "else", and "if" distinctly)

I'd like to be able to either enforce priorities on some OR operators, or explicitly rule out look-behind/look-ahead matches without capturing those ruled-out match characters, ala:

    tokens ->  ("ab" || "a" || "b")          or    tokens -> ("ab" | "a" !"b" | !"a" "b") 

The best hacks I can come up with far are:
1) Apply a regular expression in advance that inserts a boundary token at all word boundaries, then modify the grammar to match boundaries. But this makes the grammar much more complex.
2) Detect and track word boundaries in advance (without changing the input string), then use a postprocessor to reject matches that don't start at a word boundary based on the l parameter.  But this only works for leading-edge boundaries. 

So far I'm thinking of building 2, and may be able to live without trailing-edge constraints. 

Thanks for nearley, and for any ideas!

**Comment by Hardmath123** [comment]
Hi @introrse! Thanks for the kind words.

The following grammar should match the right stuff for your "ab" language, though you'll obviously need a bit of postprocessor magic to desugar the output to something nicer.

```
abs -> ("ab" | "aa" | "b"):* [ab]:?
```

or, more generally, something like

```
abs -> ("ab" | [^a] | "a" [^b] | [^a] "b"):* [ab]:?
```

So, uh, theoretically, at least, there's a way. It probably more or less corresponds to you hack number 2. But it does feel like a hack.

In practice, I can't think of a place where this is a serious problem. I feel `reject` postprocessors are powerful enough if used right. Perhaps if you showed me your use-case I could either find a way to make it work, or understand your issue better.

Keep in mind that pure negative lookaheads are "impossible" in context-free grammars because if you _could_ compile a negative lookahead into a CFG, then you'd solve the CFG intersection problem which is undecidable. :-(

P.S. Your examples so far look like you're using nearley to tokenize input. Recall that when you "parse" it into a tree-like structure, a lot of the bad tokenizations (like broken `elseif`) get rejected anyway. It's one of the cooler nearley "features". :-)

**Comment by introrse** [comment]
Thanks for the quick response, @Hardmath123 !  Alas, I oversimplified the example case:  In order to accurately reflect the problem I should have put more letters in the language and made the need to parse standalone 'a's come up.  Sorry for the poor example.  Also, my set of tokens is too large and interconnected to reasonably enumerate all the *a a* combinations in the grammar, I think.  

As you suspected, I ended up using hack 2.  I wrote a post processor function to peek back into the input string and reject matches adjacent to word characters.  To overcome the leading-edge-only limitation, I modified nearley.js to also pass in the ending location of the current match.  Here's a post-processor helper I'm using:

    wordNear = function(l,l2) {
      var before=isWordChar(PARSE_INPUT[l-1]);
      var after =isWordChar(PARSE_INPUT[l2]);
      return before || after;;
    }

Then pieces of the grammar that need to end at word boundaries get this addition:

    {% function(d,l,r,l2) {
                   if (wordNear(l,l2)) return r;
       ...

Having the global reference is unpleasant, philosophically, but aside from that and the extra post-processor complexity in the grammar this seems to work well.  I can imagine it might not be suited for a general case, but these ambiguities are rare for my grammar and use case. (parsing event dates/times out of human-originated descriptions)  Thanks for your help!
